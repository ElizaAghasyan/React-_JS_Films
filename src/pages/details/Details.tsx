import { useParams} from "react-router-dom";
import {useEffect, useState} from "react";
import {detail, getSimilar} from "../../config/movieApi";
import { apiConfig } from "../../config/apiConfig";

import styles from './Details.module.scss';
import CastList from "./CastList";
import DetailVideos from "./DetailVideos";
import MovieCard from "../../components/movieCard/MovieCard";

type detailTypes = {
    id: number | string;
    title: string;
    overview: string;
    poster_path: string;
    vote_average: number;
    backdrop_path: string;
    vote_count: number;
    name: string;
    genres: {
        name: string;
        id: number
    }[];
}

const Details = (props: detailTypes) => {
    const { category } = useParams<{category: string}>();
    const { id } = useParams<{id: string}>();

    const [items, setItems] = useState<detailTypes[] | any>([]);
    const [similar, setSimilar] = useState<detailTypes[] | any>([])

    useEffect(() => {
        const getSimilarMovies = async () => {
            let  res = await getSimilar(category, id);
            setSimilar(res.data.results.slice(0, 5))
        }
        getSimilarMovies()
    }, [category, id])

    useEffect(() => {
        const getDetails = async () => {
            let result = await detail(category, id, {params: {}});
            window.scrollTo(0, 0)
            setItems(result.data);
        }
        getDetails()
    }, [category, id]);

    let background;
    items?.backdrop_path ? background = apiConfig.imageURL(items?.backdrop_path ? items?.backdrop_path : items?.poster_path)
        : background = null

    return (
        <div className={styles.details}>
            <div className={styles.background} style={{backgroundImage: `url(${background})`}}>
                <div className={styles.background_content_background} style={{backgroundImage: `url(${apiConfig.imageURL(items.backdrop_path || items.poster_path)})`}}/>
                <div className={styles.background_info}>
                    <h1>{items.title}</h1>
                    <div className={styles.background_info_genres}>
                        {
                            items.genres && items.genres.slice(0, 5).map((genre: detailTypes) => (
                                <span key={genre.id}>{genre.name}</span>
                            ))
                        }
                    </div>
                    <p className={styles.background_info_overview}>{items.overview}</p>
                    <div className={styles.castTitle}>
                        <h2>Cast</h2>
                        <CastList />
                    </div>
                </div>
            </div>
            <div className={styles.details_videoList}>
                <DetailVideos />
            </div>
            <div className={styles.details_similar}>
                {
                    similar && similar.map((item: detailTypes) => (
                        <div key={item.id} className={styles.details_similar_item}>
                            <MovieCard item={item}/>
                        </div>
                    ))
                }
            </div>
        </div>
    );
}

export default Details;
