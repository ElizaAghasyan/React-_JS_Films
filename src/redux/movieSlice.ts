import {createAsyncThunk, createSlice} from "@reduxjs/toolkit";
import serverApi, {movieType} from "../config/movieApi";

const initialState = {
    movies: {},
}

interface MovieType  {
    id: number;
    title: string;
    overview: string;
    poster_path: string;
    vote_average: number;
    backdrop_path: string;
    vote_count: number;
    name: string;
}


export const fetchAsyncMovies = createAsyncThunk('movies/fetchAsyncMovies', async () => {
    const params = {page: 1}
    let response = await serverApi.getMovieList(movieType.popular, {params})
        .catch(err => {
            console.log(err)
        })
    let result = await response?.data.results[Math.floor(Math.random() * response.data.results.length - 1)]
    return result;
});


export const movieSlice = createSlice({
    name: 'movies',
    initialState,
    reducers: {
        addMovies: (state, {payload}) => {
            state.movies = payload
        }
    },
    extraReducers: (builder) => {
        builder.addCase(fetchAsyncMovies.pending, (state, action) => {
            console.log("Pending...")
        })
        builder.addCase(fetchAsyncMovies.fulfilled, (state: {}, action) => {
            return {
                ...state, movies: action.payload
            }
        })
        builder.addCase(fetchAsyncMovies.rejected, (state, action) => {
            console.log("Pending...")
        })
    }
});

export const {addMovies} = movieSlice.actions;

export const getBannerMovie = (state: { movies: { movies: MovieType }}) => state.movies.movies;


export default movieSlice.reducer;
