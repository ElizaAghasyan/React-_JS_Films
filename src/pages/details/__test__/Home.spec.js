import Details from "../Details";
import renderer from "react-test-renderer";
import {store} from "../../../redux/store";
import {Provider} from "react-redux";
import { createMemoryHistory } from 'history';
import {Router} from "react-router-dom";

describe('Details Component', () => {

    it('Details snapshot test', () => {

        const history = createMemoryHistory();
        const route = '/:category/:id';
        history.push(route);
        const component = renderer.create(
            <Provider store={store}>
                <Router history={history}>
                    <Details />
                </Router>
            </Provider>);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

});
