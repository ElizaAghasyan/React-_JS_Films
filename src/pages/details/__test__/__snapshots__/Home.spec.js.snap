// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Component Details snapshot test 1`] = `
<div
  className="details"
>
  <div
    className="header_styles"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <div
        className="header_content"
      >
        <h1>
          Films
        </h1>
        <div>
          <input
            onChange={[Function]}
            placeholder="Search Films"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="SearchOutlinedIcon"
            focusable="false"
            style={
              Object {
                "color": "#fff",
                "fontSize": "3em",
              }
            }
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </div>
      </div>
    </a>
  </div>
  <div
    className="background"
    style={
      Object {
        "backgroundImage": "url(null)",
      }
    }
  >
    <div
      className="background_content_background"
      style={
        Object {
          "backgroundImage": "url(https:://image.tmdb.org/t/p/original/undefined)",
        }
      }
    >
       
    </div>
    <div
      className="background_info"
    >
      <h1 />
      <div
        className="background_info_genres"
      />
      <p
        className="background_info_overview"
      />
      <div
        className="castTitle"
      >
        <h2>
          Cast
        </h2>
        <div
          className="cast"
        />
      </div>
    </div>
  </div>
  <div
    className="details_videoList"
  />
  <div
    className="details_similar"
  />
</div>
`;
